jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * (C) Copyright 2018-2018, by Alexandru Valeanu and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.isomorphism;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.connectivity.ConnectivityInspector;
+import org.jgrapht.graph.SimpleGraph;
+
+import java.util.*;
+import java.util.function.Supplier;
+
+/**
+ * This is an implementation of the AHU algorithm for detecting an isomorphism between two rooted forests.
+ * Please see <a href="http://mathworld.wolfram.com/GraphIsomorphism.html">mathworld.wolfram.com</a> for a complete
+ * definition of the isomorphism problem for general graphs.
+ *
+ * <p>
+ *     The original algorithm was first presented in "Alfred V. Aho and John E. Hopcroft. 1974.
+ *     The Design and Analysis of Computer Algorithms (1st ed.). Addison-Wesley Longman Publishing Co., Inc., Boston, MA, USA."
+ * </p>
+ *
+ * <p>
+ *     This implementation runs in linear time (in the number of vertices of the input forests)
+ *     while using a linear amount of memory.
+ * </p>
+ *
+ * <p>
+ *      For an implementation that supports both unrooted and rooted trees see {@link AHUTreeIsomorphismInspector}.
+ * </p>
+ *
+ * <p>
+ *     Note: This implementation requires the input graphs to be modifiable and to have valid vertex suppliers
```

Comment:
Previously we discussed the possibility of using AsGraphUnion.  In the case of unmodifiable inputs, its performance may be better than the cost of making modifiable copies (this is certainly true in terms of memory).  And even for modifiable inputs, phantom modifications can be troublesome, e.g. if a listener is attached to the graph.

So I think we should support it, at least as an option.  (It may require some enhancement since it currently it always assumes the worst and declares the union type as having weights and multiple edges; or that may be ignorable.)

Reply:
I am not sure it works here. I need the dummy (i.e. fresh) root to perform compute a rooted isomorphism on the union trees. I agree that we can simulate the union using `AsGraphUnion` but we do not have dummy root.

Reply:
Yes, you would still need to create the dummy root.  But you would do that in a separate graph (which would also contain the dummy edges, together with the original roots).  Then create the union on top of that plus the (unmodified) original graph.

Reply:
`AsGraphUnion` approach actually works (I was betting against it the whole time :) ). And it is not significantly slower than my previous implementation, which is excellent.